upgrade to api 1.0

jmquigs opened the pull request:
This upgrades ide-haskell to atom api 1.0.  It fixes most deprecation warnings and issues with the shadow DOM.  I didn't fix any other issues that I've noticed - such as the need to disable and re-enable the package before it activates - but I will look at those next, now that I sort of know my way around.

This also upgrades to the new autocomplete API; autocomplete no longer needs to be a package dependency.

joefitzgerald commented:
Hey @chaika2013, bump! Is there something that should be done to this PR for it to be acceptable? Please comment / merge.

joefitzgerald commented on lib/complete-provider.coffee:
```diff
@@ -36,7 +35,13 @@ PREFIX_IMPORT_FUNCTIONS    = new RegExp("[\\(\\s,]+(" + REGEXP_FUNCTION_NAME_MAY
 
-class CompleteProvider extends Provider
+class Suggestion
+  constructor: (@provider, options) ->
```
You do not need to track @provider. Consider removing it, it's not used by `autocomplete-plus`.

joefitzgerald commented on lib/complete-provider.coffee:
```diff
@@ -36,7 +35,13 @@ PREFIX_IMPORT_FUNCTIONS    = new RegExp("[\\(\\s,]+(" + REGEXP_FUNCTION_NAME_MAY
 PREFIX_FUNCTION_NAME       = new RegExp("(" + REGEXP_FUNCTION_NAME_JUST + "|" + REGEXP_ALIAS_NAME_CAP + "\\." + REGEXP_FUNCTION_NAME_MAYBE + ")$")
 
-class CompleteProvider extends Provider
+class Suggestion
```
You can use an anonymous object if you would like; you certainly don't have to, this is acceptable also.

joefitzgerald commented on lib/plugin-manager.coffee:
```diff
@@ -135,11 +138,14 @@ class PluginManager
+    # can't use the atom.services.provide API to get autocomplete,
+    # because we call registerProviderForEditor for each editor window,
+    # rather than using a single provider for everything.  so, we need access to the autocomplete module.
+    acPackage = atom.packages.getActivePackage("autocomplete-plus")
```
Do not activate autocomplete-plus. The current editor is passed in the options object to your requestHandler (`options.editor`). Use the services approach instead.

joefitzgerald commented:
@chaika2013 I take that back, this PR requires work; I am making suggestions on what should be done.

jmquigs commented:
@joefitzgerald , Thanks for the feedback.  I will make the changes that you mentioned.

jmquigs commented:
@joefitzgerald, I have updated this to the new autocomplete API.

joefitzgerald commented:
Note: the provider api wiki page has been updated to reflect the new package.json / config based registration technique. This will resolve your deprecation warnings about atom.services.

joefitzgerald commented:
Also, if you need to track the current editor and cache data associated with it, you can look at FuzzyProvider for an example of how to do that.

lierdakil commented:
I think that using in-editor variable is not an ideal solution. Using a WeakMap to track editors and their respective providers would be better IMO. You may take a look at [what I did][1] if you are unsure of how to use it (disregard anything about emitter and ghc-mod and you've got yourself a minimal example)

As a side note, @chaika2013 seems to be MIA since september. I have tried to contact him directly, but it'll likely be some time until I get an answer (if I get one).

[1]: https://github.com/lierdakil/autocomplete-haskell/blob/master/lib/autocomplete-provider.coffee

lierdakil commented on lib/utils.coffee:
```diff
@@ -14,17 +26,30 @@ isHaskellSource = (fname) ->
 
+getElementsByClass = (elem,klass) ->
+  # this is pretty obtuse, but jquery doesn't seem to support inspecting the shadowRoot directly
+  # can remove the fallback once the option to disable the shadow dom is gone
```
You might want to drop jQuery altogether and use `rootElement` instead of `shadowRoot`, which will get you root element for shadow dom or otherwise. You might also consider using `querySelectorAll` instead of `getElementsByClassName`, which has semantics almost exactly like `jQuery.find`

lierdakil commented on lib/ide-haskell.coffee:
```diff
@@ -14,23 +14,29 @@ configDefaults =
   return unless _isCabalProject
 
+  # store project root on the atom project, since the utils need it
+  atom.project.cabalProjectRoot = projRoot
```
I'm not too happy with this bit, since it sets up for possible collisions with other packages. Generally, I'm against semi-globals like this.

jmquigs replied:
@lierdakil, I don't much like this either.  The problem is that the ghc-mod utility needs the project root as the current working directory, and atom doesn't have a "project root" anymore.  I see in your project that you use the first entry from getPaths(), I considered that as well, but it made me nervous to assume the first path was what I needed.  

How about having another level of namespace on this field?  i.e atom.project.ideHaskell.cabalProjectRoot.  That hopefully makes collisions much less likely.

I could also recompute it each time by scanning for the cabal project file on each invocation, and using that to determine the root.

joefitzgerald replied:
Your provider will only ever have one instance; why not store the project root there?

jmquigs replied:
That's another option.  Just requires another accessor for the singleton.  The nice thing about atom.project is that its already available.  But yeah, storing stuff on the project is ಠ_ಠ, so I'll get rid of it.

BTW, just updated the PR with the autocomplete reg fixes, still working on the WeakMap stuff.

lierdakil commented on lib/plugin-manager.coffee:
```diff
@@ -135,33 +137,47 @@ class PluginManager
+      #  console.log "do something?"
+
+    # Register the provider
+    @acRegistration = atom.services.provide('autocomplete.provider', '1.0.0', {provider:provider})
```
atom.services got deprecated in v0.177.0, so yeah... Please refer to updated autocomplete-plus API docs, as @joefitzgerald suggested.

lierdakil commented on lib/plugin-manager.coffee:
```diff
@@ -101,18 +101,20 @@ class PluginManager
+  # Observe text editors to attach controller
+  subscribeEditorController: ->
+    @disposables.add atom.workspace.observeTextEditors (editor) =>
+      editor.haskellController = new EditorControl(editor, this)
```
WeakMap would be perfect here, IMO.

lierdakil commented on lib/complete-provider.coffee:
```diff
@@ -53,11 +58,19 @@ class CompleteProvider extends Provider
+    if (@editor.haskellCompletionProvider && @editor.haskellCompletionProvider != this)
+      # shouldn't happen, but log just in case
+      console.log "weird: editor already has a haskell completion provider"
+      @editor.haskellCompletionProvider.dispose()
```
It's a little trickier to use WeakMap here, but basically, since PluginManager is a singleton, you might maintain it there and use some sort of accessor like e.g.
```coffeescript
providerForEditor: (editor) =>
    @editorMap.get(editor)
```

joefitzgerald replied:
Don't have a provider per editor :)

See [here](https://github.com/atom-community/autocomplete-plus/blob/074a1a7a49bf806a0e830432ac997c57a2c1d08a/lib/fuzzy-provider.coffee#L16), [here](https://github.com/atom-community/autocomplete-plus/blob/074a1a7a49bf806a0e830432ac997c57a2c1d08a/lib/fuzzy-provider.coffee#L10-L11) and [here](https://github.com/atom-community/autocomplete-plus/blob/074a1a7a49bf806a0e830432ac997c57a2c1d08a/lib/fuzzy-provider.coffee#L22-L42) for how FuzzyProvider does this.

lierdakil replied:
FuzzyProvider seemingly rebuilds word list every time pane item is changed, does it not? While that's OK for simple cases, running ghc-mod is actually quite expensive both in terms of memory and cpu time. So I'd prefer having a caching provider instance for each editor rather than re-run ghc-mod every time. Alternative would be to store *cache* in a WeakMap inside (singleton) provider instance. That's also acceptable. But it's basically the same thing, and I think requires a little more boilerplate (but I might be wrong)

joefitzgerald replied:
Gotcha. Well then yeah, have a single provider and a cache, and compare options.editor with the cache. Weak map seems nice, cos it would drop the cache of the editor is destroyed.

jmquigs replied:
Having multiple completion providers allows the package to watch all open files.  A straightforward conversion of the code to a single provider just watches the currently active editor file; however, this introduces a bug; if you open a file (say "Module.hs"), make a change that should reflect in completion (like rename an exported module function), then open another file ("Main.hs") and use Atom's "Save All" command; the first file will save, but its completion database won't be rebuilt, because the provider isn't watching that file now.  The second file ("Main.hs") will then see the old name until you manually go back into the first file and save it while it is open in the editor, which forces the completion database to rebuild.  Suggestions on how to handle this?

lierdakil replied:
Could you just subscribe provider to all open editors (or rather, buffers)? `Buffer.onDidSave` and `Buffer.onWillSave` pass an `Object {path: uri}` as a first argument to callbacks, so `@currentBuffer` is not really needed anyway.
onSaved could be rewritten like this, for example:
```coffeescript
onSaved: ({path}) =>
   fileName=path
   ...
```

joefitzgerald replied:
You're on the right track there... just make sure you clean up when the buffer goes away.

joefitzgerald commented:
@jmquigs Providers are registered for scopes. Scopes have no notion of filename. So you really need to move to a cache (which _is_ file aware) within a single provider.

That is, you would track all open editors, in a single provider. You would cache the results of your long-running activities (by file). You would attach editor events such that when a file is saved, you update the cache for that file (if changed).

joefitzgerald commented:
By the way, I am planning on completely removing the old API on Tuesday, February 10th, 2015. I will obviously closely watch this issue in the interim, as you're clearly making lots of progress and I don't want to cause issues if you're 99% of the way to getting this solved.

lierdakil commented:
@joefitzgerald, well, I hear that ide-haskell autocompletion does not work anyway at the moment (#35 and also [this][1]), so it's not like you'll break anything by removing old ac-plus API.

[1]:https://github.com/lierdakil/haskell-ghc-mod/issues/1#issuecomment-72316232

lierdakil commented:
@joefitzgerald, as for your previous comment, there is slight miscommunication here, I reckon. There is a provider instance one passes to autocomplete-plus via service API. And then there's *suggestion builder*, which is called `CompletionProvider` here for historical reasons. Second one *can* and in this case *has* to be buffer-aware and whatnot.

jmquigs commented:
@lierdakil is right, the class called "CompleteProvider" here is an internal per-editor object that is not registered with autocomplete.  It _used_ to be registered with autocomplete.  I probably should have renamed it when I stopped registering it; "SuggestionBuilder" is a good name.  The actual provider is registered in the new "provideAutocomplete" function which I just added earlier today, and that is a singleton.

That being said, I probably can get away with a single CompleteProvider/SuggestionBuilder instead of multiple, and just track the list open files separately as @lierdakil mentioned.  Just want to be careful that I don't break stuff further here; there isn't much in the way of specs here (yet)...

Also Re: feb 10, yeah the deployed ide-haskell plugin is pretty much broken already (even beyond autocomplete issues).  There are actually some bugs that I want to fix with it still that this PR doesn't address.

lierdakil commented:
@jmquigs please bear in mind that further bug-fixing may be postponed for months. I managed to convince @chaika2013 to merge and deploy this PR when ready (assuming it will be ready soon), but he's extremely busy right now, so I seriously doubt he'll find time to deploy further work for a while.

TL;DR we need to solve package-breaking bugs in this PR, and pray it would last.

joefitzgerald commented:
So you guys have reached out to the owner of this repo and there's no ETA on when he'll be back to maintain? If so, might want to pursue transferring this to an organization (or forking it to one) like atom-community, so you can add committeers and not get stuck like this.

Or am I reading this wrong and the owner is willing to merge when this is done?

lierdakil commented:
@joefitzgerald For your question, see my previous comment. TL;DR -- he is willing to merge this as soon as possible, but not anything beyond this (for now).

Btw, it's a bit on offtopic, but could you enlighten me as to how package ownership works at atom.io? Specifically, is it possible to give someone permissions to publish updates without sharing credentials and/or API keys?

jmquigs commented:
Yes, ideally we'd be able to deploy updates to the package; it sounds like he will do an update for this PR, but after that, perhaps not.  

As for this PR, the only big issue that I haven't addressed yet is the fact that this package still requires you to manually deactivate and then activate ide-haskell before it "kicks in" after opening a project.  Thats pretty annoying, and I haven't delved into it yet, but I don't think it will be too hard to fix.  There are a couple other minor bugs that I have noticed that don't have fixes here either.

joefitzgerald commented:
If @chaika2013 would give anyone on this thread commit rights to this repo, they could then publish package releases

joefitzgerald commented:
Sounds like @chaika2013 should take himself off the critical path if he doesn't have time and consider adding one or both of you as a committer.

lierdakil commented:
I'll discuss it with him when I have a chance.

jmquigs commented:
Some more changes:

- I didn't do the merge of all the "provider" objects into one; I started down that path and it was looking pretty gnarly.  Not opposed to it, but would prefer to have some specs there first.  Anyway, its an implementation detail since only a single provider object is registered with autocomplete, so I think it is safe to punt on this.
- It now uses WeakMaps to track controllers and completion providers.  I still need to do some manual destruction of the weakmap values, because they both require explicit disposal functions to be called.
- Not using shadowRoot anymore
- No misc state variables are being attached to atom.project or editor objects anymore.  I fixed the project issue by just having a project settings object that contains the discovered project root.
- I fixed the bug that prevents the package from activating if a cabal file is not detected on startup (yay! no more deactivating/activating manually)

@lierdakil, can you pull down to review?  I'd appreciate it.

lierdakil commented:
Looking right now.

lierdakil commented on package.json:
```diff
@@ -8,9 +8,17 @@
   "engines": {
     "atom": ">=0.116.0"
```
Please bump atom version to 0.177.0 or whatever version you are using. It's definitely not 0.116.0. I believe apm just recently implemented checking this version, but it's a good practice nonetheless.

joefitzgerald replied:
I'm pretty sure apm doesn't respect this yet. But yes, it's good hygiene.

lierdakil commented:
I'd also appreciate if you updated changelog (I reckon this should be a minor release, but that's up for debate). Other than that, it looks good to me.

jmquigs commented:
I pushed these updates, thanks for reviewing

lierdakil commented:
All right. Thanks for your work @jmquigs! Now to get this deployed...

lierdakil commented:
Right. Merged and deployed. I've also got commit permissions for this repo, so feel free to create other pull requests if you so desire.

jmquigs commented:
@lierdakil, cool!  Thanks for doing that.  There are a few other issues I'm going to look at.   They should be smaller PRs than this one.
